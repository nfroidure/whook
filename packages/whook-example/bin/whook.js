#! /usr/bin/env node

import { join } from 'node:path';
import { URL } from 'node:url';
import { runCLI } from '@whook/whook';

const { prepareCommand } = await import(
  new URL(join('..', 'src', 'index.ts'), import.meta.url)
);

await runCLI(prepareCommand);
