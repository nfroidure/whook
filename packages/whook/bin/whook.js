#! /usr/bin/env node

import path from 'path';
import runCLI from '../dist/cli.js';

const { prepareCommand } = await import(
  path.join(process.cwd(), 'dist', 'index.js')
);

await runCLI(prepareCommand);
