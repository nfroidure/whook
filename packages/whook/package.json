{
  "name": "@whook/whook",
  "version": "4.0.0-alpha.29",
  "description": "Build strong and efficient REST web services.",
  "metapak": {
    "configs": [
      "main",
      "readme",
      "eslint",
      "babel",
      "jest",
      "jsarch",
      "jsdocs",
      "typescript"
    ],
    "data": {
      "childPackage": true,
      "testsFiles": "'src/**/*.test.ts'",
      "files": "'src/**/*.ts'",
      "ignore": [
        "dist"
      ],
      "typesFiles": "src/**/*.ts",
      "typesDefs": "src/index.d.ts",
      "bundleFiles": [
        "dist/**/*.js",
        "src/index.d.ts",
        "dist",
        "src"
      ]
    }
  },
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "architecture": "jsarch 'src/**/*.ts' > ARCHITECTURE.md && git add ARCHITECTURE.md",
    "changelog": "conventional-changelog -p angular -i CHANGELOG.md -s",
    "cli": "env NODE_ENV=${NODE_ENV:-cli}",
    "compile": "babel --extensions '.ts,.js' src --out-dir=dist --source-maps=true",
    "cover": "npm run jest -- --coverage",
    "cz": "env NODE_ENV=${NODE_ENV:-cli} git cz",
    "doc": "echo \"# API\" > API.md; jsdoc2md undefined >> API.md && git add API.md",
    "jest": "NODE_ENV=test jest",
    "lint": "eslint 'src/**/*.ts'",
    "metapak": "metapak",
    "precz": "npm t && npm run lint && npm run metapak -- -s && npm run compile",
    "prettier": "prettier --write 'src/**/*.ts'",
    "preversion": "npm t && npm run lint && npm run compile && npm run metapak -- -s",
    "test": "npm run jest",
    "types": "rimraf -f 'dist/**/*.d.ts' && tsc --project . --declaration --emitDeclarationOnly --outDir dist"
  },
  "keywords": [
    "REST",
    "HTTP",
    "server",
    "web",
    "services",
    "OpenAPI"
  ],
  "author": {
    "name": "Nicolas Froidure",
    "email": "nicolas.froidure@insertafter.com",
    "url": "http://insertafter.com/en/index.html"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/nfroidure/whook.git"
  },
  "bugs": {
    "url": "https://github.com/nfroidure/whook/issues"
  },
  "homepage": "https://github.com/nfroidure/whook",
  "devDependencies": {
    "@babel/cli": "^7.8.3",
    "@babel/core": "^7.8.3",
    "@babel/plugin-proposal-class-properties": "^7.8.3",
    "@babel/plugin-proposal-object-rest-spread": "^7.8.3",
    "@babel/preset-env": "^7.8.3",
    "@babel/preset-typescript": "^7.8.3",
    "@babel/register": "^7.8.3",
    "@types/jest": "^24.9.0",
    "@typescript-eslint/eslint-plugin": "^2.16.0",
    "@typescript-eslint/parser": "^2.16.0",
    "babel-eslint": "^10.0.3",
    "babel-plugin-knifecycle": "^1.1.1",
    "eslint": "^6.8.0",
    "eslint-plugin-prettier": "^3.1.2",
    "jest": "^24.9.0",
    "jsarch": "^2.0.3",
    "jsdoc-to-markdown": "^5.0.3",
    "metapak": "^3.1.6",
    "metapak-nfroidure": "9.8.1",
    "prettier": "^1.19.1",
    "typescript": "^3.7.5"
  },
  "dependencies": {
    "@whook/http-router": "^4.0.0-alpha.29",
    "@whook/http-server": "^4.0.0-alpha.29",
    "@whook/http-transaction": "^4.0.0-alpha.29",
    "common-services": "^6.2.0",
    "debug": "^4.1.1",
    "dotenv": "^8.2.0",
    "internal-ip": "^5.0.0",
    "knifecycle": "^8.1.0",
    "openapi-types": "^1.3.5",
    "pkg-dir": "^4.2.0",
    "portfinder": "^1.0.25"
  },
  "contributors": [],
  "engines": {
    "node": ">=8.12.0"
  },
  "babel": {
    "presets": [
      "@babel/typescript",
      [
        "@babel/env",
        {
          "targets": {
            "node": "8.12.0"
          }
        }
      ]
    ],
    "plugins": [
      "@babel/proposal-class-properties",
      "@babel/plugin-proposal-object-rest-spread",
      "babel-plugin-knifecycle"
    ],
    "sourceMaps": true
  },
  "jest": {
    "coverageReporters": [
      "lcov",
      "html"
    ],
    "testPathIgnorePatterns": [
      "/node_modules/"
    ],
    "roots": [
      "<rootDir>/src"
    ]
  },
  "eslintConfig": {
    "extends": "eslint:recommended",
    "parserOptions": {
      "ecmaVersion": 2018,
      "sourceType": "module",
      "modules": true
    },
    "env": {
      "es6": true,
      "node": true,
      "jest": true,
      "mocha": true
    },
    "plugins": [
      "prettier"
    ],
    "rules": {
      "prettier/prettier": "error"
    },
    "parser": "@typescript-eslint/parser",
    "ignorePatterns": [
      "*.d.ts"
    ],
    "overrides": [
      {
        "files": [
          "*.ts"
        ],
        "rules": {
          "no-unused-vars": [
            1,
            {
              "args": "none"
            }
          ]
        }
      }
    ]
  },
  "prettier": {
    "semi": true,
    "printWidth": 80,
    "singleQuote": true,
    "trailingComma": "all",
    "proseWrap": "always"
  },
  "files": [
    "dist/**/*.js",
    "src/index.d.ts",
    "dist",
    "src",
    "LICENSE",
    "README.md",
    "CHANGELOG.md"
  ],
  "jsarch": {
    "parserOptions": {
      "plugins": [
        "typescript"
      ]
    }
  }
}
