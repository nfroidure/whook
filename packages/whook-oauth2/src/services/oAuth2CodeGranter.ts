import { autoService, location } from 'knifecycle';
import { noop } from '@whook/whook';
import { YError } from 'yerror';
import { type LogService } from 'common-services';
import {
  type OAuth2GranterService,
  type OAuth2CodeService,
  type CheckApplicationService,
} from './oAuth2Granters.js';

export const CODE_GRANTER_TYPE = 'code';

export type OAuth2CodeGranterDependencies = {
  oAuth2Code: OAuth2CodeService;
  checkApplication: CheckApplicationService;
  log?: LogService;
};
export type OAuth2CodeGranterParameters = {
  code: string;
  redirectURI: string;
  clientId: string;
};
export type OAuth2CodeGranterService = OAuth2GranterService<
  Record<string, unknown>,
  Record<string, unknown>,
  OAuth2CodeGranterParameters
>;

export default location(autoService(initOAuth2CodeGranter), import.meta.url);

// Authorization Code Grant
// https://tools.ietf.org/html/rfc6749#section-4.1
async function initOAuth2CodeGranter({
  oAuth2Code,
  checkApplication,
  log = noop,
}: OAuth2CodeGranterDependencies): Promise<OAuth2CodeGranterService> {
  const authorizeWithCode: NonNullable<
    OAuth2CodeGranterService['authorizer']
  >['authorize'] = async ({ clientId, redirectURI, scope = '' }) => {
    const { redirectURI: finalRedirectURI } = await checkApplication({
      applicationId: clientId,
      type: CODE_GRANTER_TYPE,
      redirectURI,
      scope,
    });

    return {
      applicationId: clientId,
      redirectURI: finalRedirectURI,
      scope,
    };
  };

  // Authorization Code Response:
  // https://tools.ietf.org/html/rfc6749#section-4.1.2
  const acknowledgeWithCode: NonNullable<
    OAuth2CodeGranterService['acknowledger']
  >['acknowledge'] = async (
    authenticationData,
    { clientId, redirectURI, scope },
    additionalParameters,
  ) => {
    const code = await oAuth2Code.create(
      { ...authenticationData, applicationId: clientId, scope },
      redirectURI,
      additionalParameters,
    );

    return {
      // TODO: check a way to avoid this by adding params
      ...authenticationData,
      applicationId: clientId,
      redirectURI,
      scope,
      code,
    };
  };

  const authenticateWithCode: NonNullable<
    OAuth2CodeGranterService['authenticator']
  >['authenticate'] = async (
    { code, clientId, redirectURI },
    authenticationData,
  ) => {
    // The client must be authenticated (for now, see below)
    if (!authenticationData) {
      throw new YError('E_UNAUTHORIZED');
    }

    // This check is not really necessary atm but it acts
    // as a reminder that this grant type could be used
    // without authenticating the client. In this
    // scenario, the authenticationData should be deducted
    // from the clientId and the code given in parameters
    // See https://tools.ietf.org/html/rfc6749#section-4.1.3
    if (clientId && clientId !== authenticationData.applicationId) {
      throw new YError(
        'E_UNAUTHORIZED',
        clientId,
        authenticationData.applicationId,
      );
    }

    await checkApplication({
      applicationId: authenticationData.applicationId,
      type: CODE_GRANTER_TYPE,
      scope: '',
    });

    const newAuthenticationData = await oAuth2Code.check(
      authenticationData,
      code,
      redirectURI,
    );

    return newAuthenticationData;
  };

  log('debug', 'ðŸ‘« - OAuth2CodeGranter Service Initialized!');

  return {
    type: CODE_GRANTER_TYPE,
    authorizer: {
      responseType: 'code',
      authorize: authorizeWithCode,
    },
    acknowledger: {
      acknowledgmentType: 'code',
      acknowledge: acknowledgeWithCode,
    },
    authenticator: {
      grantType: 'authorization_code',
      authenticate: authenticateWithCode,
    },
  };
}
