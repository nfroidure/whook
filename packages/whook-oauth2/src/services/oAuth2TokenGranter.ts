import { autoService, location } from 'knifecycle';
import { noop } from '@whook/whook';
import { type LogService, type TimeService } from 'common-services';
import {
  type OAuth2GranterService,
  type OAuth2AccessTokenService,
  type CheckApplicationService,
  type OAuth2CodeService,
} from './oAuth2Granters.js';

export const IMPLICIT_GRANTER_TYPE = 'token';

export type OAuth2TokenGranterDependencies = {
  oAuth2Code: OAuth2CodeService;
  checkApplication: CheckApplicationService;
  oAuth2AccessToken: OAuth2AccessTokenService;
  time?: TimeService;
  log?: LogService;
};
export type OAuth2TokenGranterParameters = Record<string, unknown>;
export type OAuth2TokenGranterService = OAuth2GranterService<
  Record<string, unknown>,
  Record<string, unknown>,
  OAuth2TokenGranterParameters
>;

export default location(autoService(initOAuth2TokenGranter), import.meta.url);

// Implicit Grant
// https://tools.ietf.org/html/rfc6749#section-4.2
async function initOAuth2TokenGranter({
  checkApplication,
  oAuth2AccessToken,
  time = Date.now.bind(Date),
  log = noop,
}: OAuth2TokenGranterDependencies): Promise<OAuth2TokenGranterService> {
  const authorizeWithToken: NonNullable<
    OAuth2TokenGranterService['authorizer']
  >['authorize'] = async ({ clientId, redirectURI, scope }) => {
    const { redirectURI: finalRedirectURI } = await checkApplication({
      applicationId: clientId,
      type: IMPLICIT_GRANTER_TYPE,
      redirectURI,
      scope,
    });

    return {
      applicationId: clientId,
      redirectURI: finalRedirectURI,
      scope,
    };
  };

  // Access Token Response:
  // https://tools.ietf.org/html/rfc6749#section-4.2.2
  const acknowledgeWithToken: NonNullable<
    OAuth2GranterService<OAuth2TokenGranterParameters>['acknowledger']
  >['acknowledge'] = async (
    authenticationData,
    { clientId, redirectURI, scope: providedScope },
    additionalParameters,
  ) => {
    const { redirectURI: finalRedirectURI } = await checkApplication({
      applicationId: clientId,
      redirectURI,
      scope: providedScope,
      type: IMPLICIT_GRANTER_TYPE,
    });

    const { token: accessToken, expiresAt: accessTokenExpiresAt } =
      await oAuth2AccessToken.create(
        authenticationData,
        {
          // TODO: check a way to avoid this by adding params
          ...authenticationData,
          applicationId: clientId,
          scope: providedScope,
        },
        additionalParameters,
      );

    return {
      // TODO: check a way to avoid this by adding params
      ...authenticationData,
      applicationId: clientId,
      redirectURI: finalRedirectURI,
      scope: providedScope,
      accessToken,
      tokenType: 'bearer',
      expiresIn: Math.round((accessTokenExpiresAt - time()) / 1000),
    };
  };

  log('debug', 'ðŸ‘« - OAuth2TokenGranter Service Initialized!');

  return {
    type: IMPLICIT_GRANTER_TYPE,
    authorizer: {
      responseType: 'token',
      authorize: authorizeWithToken,
    },
    acknowledger: {
      acknowledgmentType: 'token',
      acknowledge: acknowledgeWithToken,
    },
  };
}
