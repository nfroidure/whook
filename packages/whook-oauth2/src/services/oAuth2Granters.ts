import { initializer, location } from 'knifecycle';
import {
  DEFAULT_ERROR_URI,
  DEFAULT_HELP_URI,
  type WhookErrorsDescriptors,
} from '@whook/whook';
import { type WhookAuthenticationData } from '@whook/authorization';
import { type CodeChallengeMethod } from './oAuth2CodeGranter.js';

export const OAUTH2_ERRORS_DESCRIPTORS: WhookErrorsDescriptors = {
  E_UNKNOWN_AUTHORIZER_TYPE: {
    code: 'unsupported_response_type',
    status: 400,
    description: `The type "$0" is not supported.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_UNKNOWN_APPLICATION_ID: {
    code: 'unauthorized_client',
    status: 400,
    description: `The client id "$0" is unknown.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_UNAUTHORIZED_APPLICATION_ID: {
    code: 'unauthorized_client',
    status: 401,
    description: `The client id "$0" is not allowed.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_APPLICATION_DISABLED: {
    code: 'unauthorized_client',
    status: 401,
    description: `The client id "$0" is disabled.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_BAD_REDIRECT_URI: {
    code: 'invalid_request',
    status: 400,
    description: `The redirect uri "$0" is not allowed.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_UNAUTHORIZED_SCOPE: {
    code: 'invalid_scope',
    status: 403,
    description: `The scope "$0" is not allowed.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_UNAUTHORIZED_GRANT_TYPE: {
    code: 'invalid_request',
    status: 403,
    description: `The response type "$0" is not allowed for this application.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_PKCE_REQUIRED: {
    code: 'invalid_request',
    status: 400,
    description: 'Code challenge required',
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_PKCE_NOT_SUPPORTED: {
    code: 'invalid_request',
    status: 400,
    description: 'Code challenge not supported for that response type ($0)',
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_UNAUTHORIZED_CLIENT: {
    code: 'invalid_client',
    status: 401,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_BAD_USERNAME: {
    code: 'invalid_request',
    status: 400,
    description: 'User does not exists.',
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_BAD_PASSWORD: {
    code: 'invalid_request',
    status: 400,
    description: 'Bad password sent for that user.',
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_BAD_REFRESH_TOKEN: {
    code: 'invalid_request',
    status: 400,
    description: 'The refresh token is expired/invalid.',
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_BAD_AUTHORIZATION_CODE: {
    code: 'invalid_request',
    status: 400,
    description: 'The authorization code is invalid or expired.',
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_UNKNOWN_ACKNOWLEDGOR_TYPE: {
    code: 'unsupported_response_type',
    status: 400,
    description: `Type "$0" not supported.`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_ACCESS_DENIED: {
    code: 'access_denied',
    status: 403,
    description: 'The user denied access to your application.',
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_UNKNOWN_AUTHENTICATION_TYPE: {
    code: 'unsupported_grant_type',
    status: 400,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_OAUTH2: {
    code: 'server_error',
    status: 500,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
  E_REFRESH_COOKIE: {
    code: 'invalid_request',
    status: 400,
    description: `Could not find any refresh_token value in the cookie header ($0).`,
    uri: DEFAULT_ERROR_URI,
    help: DEFAULT_HELP_URI,
  },
};

export type OAuth2CodeService<CODE = string> = {
  create: (
    authenticationData: WhookAuthenticationData,
    redirectURI: string,
    additionalParameters: {
      codeChallenge: string;
      codeChallengeMethod: CodeChallengeMethod;
      [name: string]: unknown;
    },
  ) => Promise<CODE>;
  check: (
    authenticationData: WhookAuthenticationData,
    code: CODE,
    redirectURI: string,
    codeVerifier?: string,
  ) => Promise<
    WhookAuthenticationData & {
      redirectURI: string;
    } & { [name: string]: unknown }
  >;
};

export type OAuth2PasswordService<USERNAME = string, PASSWORD = string> = {
  check: (
    authenticationData: WhookAuthenticationData,
    username: USERNAME,
    password: PASSWORD,
    demandedScope: WhookAuthenticationData['scope'],
  ) => Promise<WhookAuthenticationData>;
};

export type OAuth2AccessTokenService<TOKEN = string> = {
  create: (
    authenticationData: WhookAuthenticationData,
    tokenAuthenticationData: WhookAuthenticationData,
    additionalParameters?: { [name: string]: unknown },
  ) => Promise<{
    token: TOKEN;
    expiresAt: number;
  }>;
  check: (
    authenticationData: WhookAuthenticationData,
    token: TOKEN,
    scope?: WhookAuthenticationData['scope'],
  ) => Promise<WhookAuthenticationData>;
};

export type OAuth2RefreshTokenService<TOKEN = string> =
  OAuth2AccessTokenService<TOKEN>;

export type CheckApplicationService = {
  (context: {
    applicationId: string;
    type: string;
    scope: string;
    redirectURI?: string;
  }): Promise<{
    applicationId: string;
    type: string;
    scope: string;
    redirectURI: string;
  }>;
};

export type OAuth2GranterAuthorize<
  AUTHORIZE_PARAMETERS extends Record<string, unknown> = Record<
    string,
    unknown
  >,
> = (
  context: {
    clientId: WhookAuthenticationData['applicationId'];
    redirectURI: string;
    scope: WhookAuthenticationData['scope'];
  },
  authorizeParameters: AUTHORIZE_PARAMETERS,
) => Promise<{
  applicationId: WhookAuthenticationData['applicationId'];
  redirectURI: string;
  scope: WhookAuthenticationData['scope'];
}>;

export type Oauth2GranterAuthenticate<
  GRANT_PARAMETERS extends Record<string, unknown> = Record<string, unknown>,
> = (
  grantParameters: GRANT_PARAMETERS,
  authenticationData: WhookAuthenticationData,
) => Promise<WhookAuthenticationData>;

export type OAuth2GranterAcknowledge<
  ACKNOWLEDGE_PARAMETERS extends Record<string, unknown> = Record<
    string,
    unknown
  >,
> = (
  authenticationData: WhookAuthenticationData,
  acknowledgeParameters: {
    clientId: WhookAuthenticationData['applicationId'];
    redirectURI: string;
    scope: WhookAuthenticationData['scope'];
  },
  additionalParameters: ACKNOWLEDGE_PARAMETERS,
) => Promise<
  WhookAuthenticationData & {
    redirectURI: string;
    [name: string]: unknown;
  }
>;

export type OAuth2GranterService<
  AUTHORIZE_PARAMETERS extends Record<string, unknown> = Record<
    string,
    unknown
  >,
  ACKNOWLEDGE_PARAMETERS extends Record<string, unknown> = Record<
    string,
    unknown
  >,
  GRANT_PARAMETERS extends Record<string, unknown> = Record<string, unknown>,
> = {
  type: string;
  authorizer?: {
    responseType: string;
    authorize: OAuth2GranterAuthorize<AUTHORIZE_PARAMETERS>;
  };
  acknowledger?: {
    acknowledgmentType: string;
    acknowledge: OAuth2GranterAcknowledge<ACKNOWLEDGE_PARAMETERS>;
  };
  authenticator?: {
    grantType: string;
    authenticate: Oauth2GranterAuthenticate<GRANT_PARAMETERS>;
  };
};

export type OAuth2Options = {
  authenticateURL: string;
  defaultToClientScope?: boolean;
  forcePKCE?: boolean;
};

export type OAuth2Config = {
  OAUTH2: OAuth2Options;
  ERRORS_DESCRIPTORS: WhookErrorsDescriptors;
};

export default location(
  initializer(
    {
      name: 'oAuth2Granters',
      type: 'service',
      inject: [
        'oAuth2ClientCredentialsGranter',
        'oAuth2CodeGranter',
        'oAuth2PasswordGranter',
        'oAuth2RefreshTokenGranter',
        'oAuth2TokenGranter',
      ],
    },
    async (services) => Object.keys(services).map((key) => services[key]),
  ),
  import.meta.url,
);
