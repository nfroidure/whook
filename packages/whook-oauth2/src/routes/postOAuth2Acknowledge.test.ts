import { describe, test, beforeEach, jest, expect } from '@jest/globals';
import initPostOAuth2Acknowledge from './postOAuth2Acknowledge.js';
import { YError } from 'yerror';
import { DEFAULT_ERRORS_DESCRIPTORS } from '@whook/whook';
import { OAUTH2_ERRORS_DESCRIPTORS } from '../services/oAuth2Granters.js';
import { type CheckApplicationService } from '../services/oAuth2Granters.js';
import { type LogService } from 'common-services';
import { type OAuth2GranterService } from '../services/oAuth2Granters.js';
import { type OAuth2CodeGranterService } from '../services/oAuth2CodeGranter.js';
import { type OAuth2TokenGranterService } from '../services/oAuth2TokenGranter.js';

describe('postOAuth2Acknowledge', () => {
  const ERRORS_DESCRIPTORS = {
    ...DEFAULT_ERRORS_DESCRIPTORS,
    ...OAUTH2_ERRORS_DESCRIPTORS,
  };
  const checkApplication = jest.fn<CheckApplicationService>();
  const log = jest.fn<LogService>();
  const codeGranter = {
    type: 'code',
    authorizer: {
      responseType: 'code',
      authorize:
        jest.fn<
          NonNullable<OAuth2CodeGranterService['authorizer']>['authorize']
        >(),
    },
    acknowledger: {
      acknowledgmentType: 'code',
      acknowledge:
        jest.fn<
          NonNullable<OAuth2CodeGranterService['acknowledger']>['acknowledge']
        >(),
    },
    authenticator: {
      grantType: 'authorization_code',
      authenticate:
        jest.fn<
          NonNullable<OAuth2CodeGranterService['authenticator']>['authenticate']
        >(),
    },
  };
  const tokenGranter = {
    type: 'token',
    authorizer: {
      responseType: 'token',
      authorize:
        jest.fn<
          NonNullable<
            OAuth2TokenGranterService['authenticator']
          >['authenticate']
        >(),
    },
    acknowledger: {
      acknowledgmentType: 'token',
      acknowledge:
        jest.fn<
          NonNullable<OAuth2TokenGranterService['acknowledger']>['acknowledge']
        >(),
    },
  };
  const oAuth2Granters = [
    codeGranter,
    tokenGranter,
  ] as unknown as OAuth2GranterService<
    Record<string, unknown>,
    Record<string, unknown>,
    Record<string, unknown>
  >[];

  beforeEach(() => {
    checkApplication.mockReset();
    log.mockReset();
    [
      codeGranter.authorizer.authorize,
      codeGranter.acknowledger.acknowledge,
      codeGranter.authenticator.authenticate,
      tokenGranter.authorizer.authorize,
      tokenGranter.acknowledger.acknowledge,
    ].forEach((mock) => mock.mockReset());
  });

  test('should redirect', async () => {
    [
      codeGranter.authorizer.authorize,
      codeGranter.authenticator.authenticate,
      tokenGranter.authorizer.authorize,
      tokenGranter.acknowledger.acknowledge,
    ].forEach((mock) =>
      mock.mockRejectedValueOnce(new YError('E_NOT_SUPPOSED_TO_BE_HERE')),
    );
    codeGranter.acknowledger.acknowledge.mockResolvedValueOnce({
      applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
      redirectURI: 'http://lol',
      scope: 'user',
      userId: '1',
    });

    const postOAuth2Acknowledge = await initPostOAuth2Acknowledge({
      ERRORS_DESCRIPTORS,
      oAuth2Granters,
      checkApplication,
      log,
    });
    const response = await postOAuth2Acknowledge({
      authenticationData: {
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        scope: 'auth',
        userId: '1',
      },
      body: {
        responseType: 'code',
        clientId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        redirectURI: 'https://www.example.com',
        scope: 'user',
        state: 'bancal',
        acknowledged: true,
      },
    });
    expect({
      response,
    }).toMatchInlineSnapshot(`
      {
        "response": {
          "headers": {
            "location": "https://www.example.com/?client_id=abbacaca-abba-caca-abba-cacaabbacaca&scope=user&state=bancal&redirect_uri=http%3A%2F%2Flol&user_id=1",
          },
          "status": 302,
        },
      }
    `);
    expect({
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      checkApplicationCalls: checkApplication.mock.calls,
      codeGranterAuthorizerAuthorizeCalls:
        codeGranter.authorizer.authorize.mock.calls,
      codeGranterAcknowledgerAcknowledgeCalls:
        codeGranter.acknowledger.acknowledge.mock.calls,
      codeGranterAuthenticatorAuthenticateCalls:
        codeGranter.authenticator.authenticate.mock.calls,
      tokenGranterAuthorizerAuthorizeCalls:
        tokenGranter.authorizer.authorize.mock.calls,
      tokenGranterAcknowledgerAcknowledgeCalls:
        tokenGranter.acknowledger.acknowledge.mock.calls,
    }).toMatchSnapshot();
  });

  test('should redirect errors too', async () => {
    [
      codeGranter.authorizer.authorize,
      codeGranter.authenticator.authenticate,
      tokenGranter.authorizer.authorize,
      tokenGranter.acknowledger.acknowledge,
      codeGranter.acknowledger.acknowledge,
    ].forEach((mock) =>
      mock.mockRejectedValueOnce(new YError('E_NOT_SUPPOSED_TO_BE_HERE')),
    );

    const postOAuth2Acknowledge = await initPostOAuth2Acknowledge({
      ERRORS_DESCRIPTORS,
      oAuth2Granters,
      checkApplication,
      log,
    });
    const response = await postOAuth2Acknowledge({
      authenticationData: {
        applicationId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        scope: 'auth',
        userId: '1',
      },
      body: {
        responseType: 'yolo',
        clientId: 'abbacaca-abba-caca-abba-cacaabbacaca',
        redirectURI: 'https://www.example.com',
        scope: 'user',
        state: 'bancal',
        acknowledged: true,
      },
    });
    expect({
      response,
    }).toMatchInlineSnapshot(`
      {
        "response": {
          "headers": {
            "location": "https://www.example.com/?error=unsupported_response_type&error_decription=Type+%22yolo%22+not+supported.",
          },
          "status": 302,
        },
      }
    `);
    expect({
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
      checkApplicationCalls: checkApplication.mock.calls,
      codeGranterAuthorizerAuthorizeCalls:
        codeGranter.authorizer.authorize.mock.calls,
      codeGranterAcknowledgerAcknowledgeCalls:
        codeGranter.acknowledger.acknowledge.mock.calls,
      codeGranterAuthenticatorAuthenticateCalls:
        codeGranter.authenticator.authenticate.mock.calls,
      tokenGranterAuthorizerAuthorizeCalls:
        tokenGranter.authorizer.authorize.mock.calls,
      tokenGranterAcknowledgerAcknowledgeCalls:
        tokenGranter.acknowledger.acknowledge.mock.calls,
    }).toMatchSnapshot();
  });
});
