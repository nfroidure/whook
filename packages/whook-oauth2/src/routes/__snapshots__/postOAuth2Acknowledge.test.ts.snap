// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`postOAuth2Acknowledge should redirect 2`] = `
{
  "checkApplicationCalls": [
    [
      {
        "applicationId": "abbacaca-abba-caca-abba-cacaabbacaca",
        "scope": "",
        "type": "root",
      },
    ],
  ],
  "codeGranterAcknowledgerAcknowledgeCalls": [
    [
      {
        "applicationId": "abbacaca-abba-caca-abba-cacaabbacaca",
        "scope": "auth",
        "userId": "1",
      },
      {
        "clientId": "abbacaca-abba-caca-abba-cacaabbacaca",
        "redirectURI": "https://www.example.com",
        "scope": "user",
      },
      {},
    ],
  ],
  "codeGranterAuthenticatorAuthenticateCalls": [],
  "codeGranterAuthorizerAuthorizeCalls": [],
  "logCalls": [],
  "tokenGranterAcknowledgerAcknowledgeCalls": [],
  "tokenGranterAuthorizerAuthorizeCalls": [],
}
`;

exports[`postOAuth2Acknowledge should redirect errors too 2`] = `
{
  "checkApplicationCalls": [
    [
      {
        "applicationId": "abbacaca-abba-caca-abba-cacaabbacaca",
        "scope": "",
        "type": "root",
      },
    ],
  ],
  "codeGranterAcknowledgerAcknowledgeCalls": [],
  "codeGranterAuthenticatorAuthenticateCalls": [],
  "codeGranterAuthorizerAuthorizeCalls": [],
  "logCalls": [
    [
      "debug",
      "ðŸ‘« - OAuth2 acknowledge error",
      "E_UNKNOWN_ACKNOWLEDGOR_TYPE",
    ],
  ],
  "tokenGranterAcknowledgerAcknowledgeCalls": [],
  "tokenGranterAuthorizerAuthorizeCalls": [],
}
`;
